roachprod-microbench: post github issues

herkolategan opened the pull request:
This PR adds functionality to automatically post GitHub issues for microbenchmark failures in CI (teamcity).

It will only post failures for executions from the master branch. Typically two binaries are interleaved to produce a comparison, but with this change the binary from master (experiment) will be executed first, if it fails it will cancel all further executions of the benchmark and post the failure to github.

There is one caveat regarding the posted issues. We retrieve stdout and stderr separately, so the log (on a github issue) will be a concatenation of the two.

Failures will be labeled with `O-microbench` as well as `C-test-failure` and `release-blocker`
A timeout (> 20 minutes for one iteration) will also count as a failure.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143738)
<!-- Reviewable:end -->

herkolategan committed:
githubpost: add microbenchmark failure type

This change adds a new type of failure that can be created by the microbenchmark
CI job.

Epic: None
Release note: None

herkolategan committed:
roachprod-microbench: set group id for commands

Previously, commands were not assigned a group id, so they were not grouped
together in the remote execution. This commit sets the group id for each command
to the benchmark name. This causes a group to be cancelled if any command in the
group fails.

Epic: None
Release note: None

herkolategan committed:
roachprod-microbench: add issue posting to github

Previously, errors were only reported to the logs (artifacts directory) and not
posted to github. This change adds a new flag, `--post-issues`, which causes the
microbenchmark CI job to post issues to github for any failed benchmarks for a
selected binary.

It uses environment variables to determine for which binary (experiment or
baseline), branch and SHA to post issues.

Epic: None
Release note: None

herkolategan committed:
build: enable github posting on microbenchmark weekly

Epic: None
Release note: None

DarrylWong commented on pkg/cmd/roachprod-microbench/executor.go:
```diff
@@ -265,6 +287,21 @@ func (e *executor) generateBenchmarkCommands(
 	binaryKeys := maps.Keys(e.binaries)
 	sort.Strings(binaryKeys)
 
+	// If post issues is enabled, move the post config binary key to the front
```
Is there any concern that there is now a very slight difference between runs with posting vs runs without? Maybe the argument is that there's a bigger difference in machines used anyway - I'd buy that.

herkolategan replied:
This could probably be a default since it shouldn't affect the results.

srosenberg commented on pkg/cmd/roachprod-microbench/executor.go:
```diff
@@ -114,14 +135,15 @@ func newExecutor(config executorConfig) (*executor, error) {
 		binaries:     map[string]string{"experiment": "experiment"},
 		outputDir:    "artifacts/roachprod-microbench",
-		timeout:      "10m",
+		timeout:      "20m",
```
Why the increase? Are there legit. benchmarks that need more than 10m?

herkolategan replied:
Nice catch, this was an unintentional change! Can't remember exactly what I was testing, but it shouldn't have made it in.

srosenberg commented on pkg/cmd/roachprod-microbench/executor.go:
```diff
@@ -265,6 +287,21 @@ func (e *executor) generateBenchmarkCommands(
+			if key == e.postConfig.binary {
+				// Move the key to front by removing it and inserting at index 0
+				binaryKeys = slices.Delete(binaryKeys, i, i+1)
+				binaryKeys = append([]string{e.postConfig.binary}, binaryKeys...)
```
Nit: equivalent `copy(binaryKeys[1:i+1], binaryKeys[0:i]); binaryKeys[0] = e.postConfig.binary` has 0 allocs

srosenberg commented on pkg/cmd/roachprod-microbench/executor.go:
```diff
@@ -410,13 +448,24 @@ func (e *executor) executeBenchmarks() error {
+				formatter, req := createBenchmarkPostRequest(artifactsDir, response, timeout)
+				err = postBenchmarkIssue(context.Background(), e.log, formatter, req)
+				if err != nil {
+					e.log.Errorf("Failed to post benchmark issue - %v", err)
```
This is effectively a silent error since it won't fail build [1].

[1] https://github.com/cockroachdb/cockroach/issues/147116

srosenberg replied:
Unless we return an error from `executeBenchmarks` when `errorCount > 0`? Except the counter doesn't appear to be used.

herkolategan replied:
Good point! We don't fail execution when benchmarks fail (anymore).
But if a post error happens I think it would be good to fail the build so there's visibility. I'll change this before merging.

herkolategan commented on build/teamcity/cockroach/nightlies/microbenchmark_weekly.sh:
```diff
@@ -139,6 +139,13 @@ for sha in "${build_sha_arr[@]}"; do
 done
 
+# Post issues to github for triggered builds (triggered builds are always on master)
+if [ -n "${TRIGGERED_BUILD:-}" ]; then
```
Correct, it's coming from the schedule, but I unscheduled it a few weeks ago due to a failure I was busy fixing.
Could use a different env var name, but generally TRIGGERED would mean: post issues, post results, post to slack etc.

herkolategan committed:
roachprod-microbench: remove unused error counter

Epic: None
Release note: None
